src/Stratis.Features.FederatedPeg/SourceChain/MaturedBlocksProvider.cs

```diff
@@ -72,17 +72,28 @@ public SerializableResult<List<MaturedBlockDepositsModel>> RetrieveDeposits(int
 
             var result = new SerializableResult<List<MaturedBlockDepositsModel>>
             {
-                Value = new List<MaturedBlockDepositsModel>()
+                Value = new List<MaturedBlockDepositsModel>(),
+                Message = ""
             };
-            
+
+            // If we're asked for blocks beyond the tip then let the caller know that there are no new blocks available.
+            if (maturityHeight > this.consensusManager.Tip.Height)
+                return result;
+
             int maxConfirmations = this.retrievalTypeConfirmations.Values.Max();
             int startHeight = maturityHeight - maxConfirmations;
-            ChainedHeader firstBlock = this.consensusManager.Tip.GetAncestor(maturityHeight);
-            firstBlock = firstBlock.EnumerateToGenesis().SkipWhile(h => h.Height > startHeight && (!this.deposits.TryGetValue(h.Height, out BlockDeposits blockDeposits) || blockDeposits.BlockHash != h.HashBlock)).FirstOrDefault();
+
+            // Determine the first block to extract deposits for.
+            ChainedHeader firstToProcess = this.consensusManager.Tip.GetAncestor(maturityHeight);
+            for (ChainedHeader verifyBlock = firstToProcess?.Previous; verifyBlock != null && verifyBlock.Height >= startHeight; verifyBlock = verifyBlock.Previous)
+                if (!this.deposits.TryGetValue(verifyBlock.Height, out BlockDeposits blockDeposits) || blockDeposits.BlockHash != verifyBlock.HashBlock)
```

Comment:
I stepped through the code a bit with one of the tests and I just want to confirm, does `deposits` now basically act like a "cache"? 

Because on first run, it looks like it doesn't matter what `maturityHeight` we pass into the method, it will always start from genesis.

So on second run, it will start from 5?

If this is the case I think we should rename deposits to `depositCache` or `processedDeposits` ?

Reply:
Yes. Let me do that in the next PR. CI takes ages to complete after every change.

Reply:
> Because on first run, it looks like it doesn't matter what maturityHeight we pass into the method, it will always start from genesis.

Only if maturityHeight is within the first maxConfirmations blocks.